### PR TITLE
5735 historical tc touchpoint loaders

### PR DESCRIPTION
**Pull request by ttoomey**
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

[Ref issue 5735](https://github.com/open-path/Green-River/issues/5735)
Historical assessment and service importers

## Type of change
[//]: # 'remove options that are not relevant'
- [x] New feature (adds functionality)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/aum_prevention_screening_loader.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    def filename
+      'AumPrevention.xlsx'
+    end
+
+    CDED_CONFIGS = [
+      { label: 'Do you live within Arlington city limits?', key: 'within_arlington', repeats: false, field_type: 'boolean' },
+      { label: 'Do you receive Section 8 or other housing subsidy?', key: 'section_8', repeats: false, field_type: 'boolean' },
+      { label: 'What size apartment do you have?', key: 'apartment_size', repeats: false, field_type: 'string' },
+      { label: 'Is your rent less that $1,064 for one bedroom?', key: 'rent_less_than_x_one_bedroom', repeats: false, field_type: 'boolean' },
+      { label: 'Is your rent less thatn $1,269 for a two bedroom?', key: 'rent_less_than_x_two_bedroom', repeats: false, field_type: 'boolean' },
```
Comment (ttoomey) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/aum_prevention_screening_loader.rb:20]: typos are intentional, match typos in export label

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_assessment_loader.rb [hunk 2]
```diff
@@ -55,12 +55,12 @@ def ce_assessment_level
     def validate_cde_configs
       seen_element_ids = Set.new
 
-      required_keys = [:label, :key, :repeats, :field_type]
-      all_keys = (required_keys + [:element_id, :ignore_type]).to_set
+      required_keys = [:key, :repeats, :field_type]
+      all_keys = (required_keys + [:label, :element_id, :ignore_type]).to_set
```
Comment (ttoomey) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_assessment_loader.rb:59]: allows the config to specify either label or element (allow optional label)

**Review by ttoomey — COMMENTED** [review]
Loaders run against local db. Have not tested against production

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_services_loader.rb [hunk 3]
```diff
@@ -235,6 +235,114 @@ def row_enrollment_id(row)
...
+              key: 'service_contact_location',
+            },
+            'Time Spent' => {
+              key: 'service_time_spent',
+              field_type: :integer,
+            },
+          },
+        },
+        'When We Love' => {
+          service_type: 'TB and Bus Pass',
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_services_loader.rb]:
```suggestion
          service_type: 'Financial Assistance',
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/dcsf_v3_loader.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+module HmisExternalApis::TcHmis::Importers::Loaders
+  class DcsfV3Loader < CustomAssessmentLoader
+    ASSESSMENT_DATE_COL = 'Date Taken'.freeze
+
+    def filename
+      'DCSF_V3.xlsx'
+    end
+
+    CDED_CONFIGS = [
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/dcsf_v3_loader.rb:15]: @ttoomey can you update these keys to match the keys that are currently referenced by the forms in staging? 

```
irb(main):039:0> Hmis::Hud::CustomDataElementDefinition.where("key LIKE ?", "dcsf_%").pluck(:key, :field_type, :label)
=>
[["dcsf_v3_agency_name", "string", "Agency Name"],
 ["dcsf_v3_funding_type", "string", "Funding Type"],
 ["dcsf_v3_funding_categories", "string", "Funding Categories"],
 ["dcsf_v3_document_type", "string", "Which type of critical document is being requested?"],
 ["dcsf_v3_rrh_or_psh", "boolean", "Is client in a Rapid Rehousing or PSH Program?"],
 ["dcsf_v3_admin_app_fee", "float", "Administration/application fee"],
 ["dcsf_v3_high_risk_opportunity_fee", "float", "High Risk/Opportunity fee"],
 ["dcsf_v3_deposit", "float", "Deposit"],
 ["dcsf_v3_pet_deposit", "float", "Pet Deposit"],
 ["dcsf_v3_damages", "float", "Damages"],
 ["dcsf_v3_utility_deposit", "float", "Utility Deposit"],
 ["dcsf_v3_bus_or_transportation", "float", "Bus Ticket or Transportation for Diversion Only"],
 ["dcsf_v3_rent_one_time_assistance", "float", "Rent (one-time assistance)"],
 ["dcsf_v3_family_motel", "float", "Family Motel"],
 ["dcsf_v3_family_day_sheltering", "float", "Family day sheltering"],
 ["dcsf_v3_additional_furniture_mattress", "float", "Additional (furniture/mattress)"],
 ["dcsf_v3_drivers_license_state_id", "float", "Driver's License/State ID"],
 ["dcsf_v3_birth_certificate", "float", "Birth Certificate"],
 ["dcsf_v3_social_security_card", "float", "Social Security Card"],
 ["dcsf_v3_marriage_death_certificate", "float", "Marriage/Death Certificate"],
 ["dcsf_v3_housing_search_transportation", "float", "Housing Search Transportation"],
 ["dcsf_v3_date_payment_made", "date", "Date payment was made"],
 ["dcsf_v3_payment_amount", "float", "Payment Amount"],
 ["dcsf_v3_name_of_apartment_complex_property_owner", "string", "Name of Apartment Complex/Property Owner"],
 ["dcsf_v3_other_efforts", "string", "What efforts have been made to secure funds other than DCSF?"],
 ["dcsf_v3_reason_for_denial", "string", "Reason for denial"],
 ["dcsf_v3_reason_for_pending_status", "string", "Reason for pending status"],
 ["dcsf_v3_approved_date", "date", "Approved Date"],
 ["dcsf_v3_final_funding_category", "string", "Final Funding Category (based on information received)"],
 ["dcsf_v3_funding_source", "string", "Funding Source"],
 ["dcsf_v3_household_type", "string", "Household Type"],
 ["dcsf_v3_status", "string", "Status"]]
```

Comment (gigxz) [reply drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/dcsf_v3_loader.rb:15]: the labels dont have to match of course, just included them for reference

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/direction_homes_housing_loader.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    def filename
+      'DirectionHomes.xlsx'
+    end
+
+    CDED_CONFIGS = [
+      { label: 'Date Application Received', key: 'date_app_received', repeats: false, field_type: 'date' },
+      { label: 'Date Application Completed', key: 'date_app_completed', repeats: false, field_type: 'date' },
+      { label: 'Date Given Voucher and Briefing Date', key: 'date_given_voucher', repeats: false, field_type: 'date' },
+      { label: 'Passed Inspection Date', key: 'date_passed_inspection', repeats: false, field_type: 'date' },
+      { label: 'Move in Date', key: 'move_in_date', repeats: false, field_type: 'date' },
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/direction_homes_housing_loader.rb:20]:
```
[["direction_homes_housing_date_app_received", "date", "Date Application Received"],
 ["direction_homes_housing_date_given_voucher", "date", "Date Given Voucher and Briefing"],
 ["direction_homes_housing_date_app_completed", "date", "Date Application Completed"],
 ["direction_homes_housing_date_passed_inspection", "date", "Passed Inspection Date"],
 ["direction_homes_housing_move_in_date", "date", "Move in Date"]]
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/document_ready_loader.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      { label: 'If outreach choose one', key: 'if_outreach_choose', repeats: false, field_type: 'string' },
+      { label: 'Are all required documents uploaded into ETO?', key: 'all_required_documents_uploaded', repeats: false, field_type: 'boolean' },
+      { label: 'Which documents are missing?', key: 'missing', repeats: true, field_type: 'string' },
+      { label: 'Case Notes.', key: 'case_notes', repeats: false, field_type: 'string' },
+      { label: "If Yes (Navigator's Name)", key: 'navigator_name', repeats: false, field_type: 'string' },
+      { label: 'Status', key: 'status', repeats: false, field_type: 'string' },
+      { label: 'Staff adding.', key: 'staff_name', repeats: false, field_type: 'string' },
+      { label: 'Status Notes', key: 'status_notes', repeats: false, field_type: 'string' },
+      { label: 'Service Provider', key: 'service_provider', repeats: false, field_type: 'string' },
+      { label: 'Assessor', key: 'assessor', repeats: false, field_type: 'string' },
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/document_ready_loader.rb:29]:
```
[["document_ready_assigned_navigator", "boolean", "Is the participant assigned to a navigator?"],
 ["document_ready_navigator_name", "string", "Navigator's Name"],
 ["document_ready_if_no_navigator", "string", "If No"],
 ["document_ready_service_provider", "string", "Service Provider"],
 ["document_ready_assessor", "string", "Assessor"],
 ["document_ready_if_shelter_choose", "string", "If shelter choose one"],
 ["document_ready_if_outreach_choose", "string", "If outreach choose one"],
 ["document_ready_all_required_documents_uploaded", "boolean", "Are all required documents uploaded?"],
 ["document_ready_missing", "string", "Which documents are missing?"],
 ["document_ready_staff_name", "string", "Staff Name"],
 ["document_ready_household_type", "string", "Household Type"],
 ["document_ready_case_notes", "string", "Case Notes"],
 ["document_ready_status_notes", "string", "Status Notes"],
 ["document_ready_status", "string", "Status"]]
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/hvrp_employment_details_loader.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      { label: 'State', key: 'employer_address_state', repeats: false, field_type: 'string' },
+      { label: 'County', key: 'employer_address_county', repeats: false, field_type: 'string' },
+      { label: 'Zip Code', key: 'employer_address_zip', repeats: false, field_type: 'string' },
+      { label: 'Current Job Title:', key: 'current_job_title', repeats: false, field_type: 'string' },
+      { label: 'Date of Hire:', key: 'date_of_hire', repeats: false, field_type: 'date' },
+      { label: 'Last Date of Employment:', key: 'last_date_of_employment', repeats: false, field_type: 'date' },
+      { label: 'Rate of Pay:', key: 'rate_of_pay', repeats: false, field_type: 'float' },
+      { label: 'Wages Change Reason (if applicable):', key: 'wages_change_reason', repeats: false, field_type: 'string' },
+      { label: 'How many hours work per week?', key: 'hours_per_week', repeats: false, field_type: 'integer' },
+      { label: 'Monthly Earnings', key: 'monthly_earnings', repeats: false, field_type: 'float' },
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/hvrp_employment_details_loader.rb:31]:
```
[["hvrp_employment_details_form_type", "string", "Employment Form Type"],
 ["hvrp_employment_details_employer_address_name", "string", "Name"],
 ["hvrp_employment_details_employer_address_company", "string", "Company"],
 ["hvrp_employment_details_employer_address_1", "string", "Address Line 1"],
 ["hvrp_employment_details_employer_address_2", "string", "Address Line 2"],
 ["hvrp_employment_details_employer_address_city", "string", "City"],
 ["hvrp_employment_details_employer_address_county", "string", "County"],
 ["hvrp_employment_details_employer_address_state", "string", "State"],
 ["hvrp_employment_details_employer_address_zip", "string", "Zip Code"],
 ["hvrp_employment_details_wages_change_reason", "string", "Wages Change Reason (if applicable)"],
 ["hvrp_employment_details_rate_of_pay", "float", "Rate of Pay"],
 ["hvrp_employment_details_hours_per_week", "integer", "How many hours work per week?"],
 ["hvrp_employment_details_monthly_earnings", "float", "Monthly Earnings"],
 ["hvrp_employment_details_date_of_hire", "date", "Date of Hire"],
 ["hvrp_employment_details_current_job_title", "string", "Current Job Title"],
 ["hvrp_employment_details_last_date_of_employment", "date", "Last Date of Employment"]]
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/hvrp_employment_details_loader.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      parse_date(row.field_value(ASSESSMENT_DATE_COL))
+    end
+
+    def row_assessment_id(row)
+      response_id = row.field_value(RESPONSE_ID_COL)
+      "hvrp_employment_details-eto-#{response_id}"
+    end
+
+    def form_definition_identifier
+      'hvrp_employment_details'
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/hvrp_employment_details_loader.rb]:
```suggestion
      'hvrp-employment-details'
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/document_ready_loader.rb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      parse_date(row.field_value(ASSESSMENT_DATE_COL))
+    end
+
+    def row_assessment_id(row)
+      response_id = row.field_value(RESPONSE_ID_COL)
+      "document_ready-eto-#{response_id}"
+    end
+
+    def form_definition_identifier
+      'document_ready'
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/document_ready_loader.rb]:
```suggestion
      'document-ready'
```

Comment (gigxz) [reply drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/document_ready_loader.rb]: these are already created. it's easier to change the expected identifier here than it is to update the forms and their rules

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/direction_homes_housing_loader.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+      parse_date(row.field_value(ASSESSMENT_DATE_COL))
+    end
+
+    def row_assessment_id(row)
+      response_id = row.field_value(RESPONSE_ID_COL)
+      "direction_homes_housing-eto-#{response_id}"
+    end
+
+    def form_definition_identifier
+      'direction_homes_housing'
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/direction_homes_housing_loader.rb]:
```suggestion
      'direction-homes-housing-application-info'
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/dcsf_v3_loader.rb [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+      parse_date(row.field_value(ASSESSMENT_DATE_COL))
+    end
+
+    def row_assessment_id(row)
+      response_id = row.field_value(RESPONSE_ID_COL)
+      "dcsf_v3-eto-#{response_id}"
+    end
+
+    def form_definition_identifier
+      'dcsf_v3'
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/dcsf_v3_loader.rb]:
```suggestion
      'dcsf-v3'
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/caseworker_exit_survey_loader.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      { label: 'One month follow-up', key: 'one_month_follow_up', repeats: true, field_type: 'string' },
+      { label: 'Referrals given at follow-up', key: 'referrals_at_follow_up', repeats: false, field_type: 'string' },
+      { label: 'If other  housing program, please specify which one.', key: 'housing_program_other', repeats: false, field_type: 'string' },
+      { label: 'Number of weeks in program', key: 'weeks_in_program', repeats: false, field_type: 'integer' },
+      { label: 'How much did income increase while in shelter?', key: 'income_increase', repeats: false, field_type: 'float' },
+      { label: 'If client exited to a housing program, which program was it?', key: 'housing_program', repeats: false, field_type: 'string' },
+      { label: 'If client left with savings, was it from employment, benefits, both or other? (choose one answer)', key: 'savings_source', repeats: false, field_type: 'string' },
+      { label: 'If other, please specify exit destination.', key: 'destination_other', repeats: false, field_type: 'string' },
+      { label: 'How many members in family?', key: 'members_in_family', repeats: false, field_type: 'integer' },
+
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/caseworker_exit_survey_loader.rb]:
```
[["caseworker_exit_date", "date", "Exit Date", false],
 ["caseworker_exit_phone", "string", "Phone Number", false],
 ["caseworker_exit_email", "string", "Email", false],
 ["caseworker_exit_weeks_in_program", "integer", "Number of weeks in program", false],
 ["caseworker_exit_beginning_gas_score", "integer", "Beginning GAS Score", false],
 ["caseworker_exit_beginning_gas_score_date", "date", "Beginning GAS Score Date", false],
 ["caseworker_exit_end_gas_score", "integer", "Exit GAS Score", false],
 ["caseworker_exit_end_gas_score_date", "date", "Exit GAS Score Date", false],
 ["caseworker_exit_gas_score_change", "integer", "Difference/Change in GAS score", false],
 ["caseworker_exit_gas_score_percent_change", "integer", "Percent of GAS Change", false],
 ["caseworker_exit_destination", "string", "Where will the client be residing?", false],
 ["caseworker_exit_destination_other", "string", "If other, please specify exit destination.", false],
 ["caseworker_exit_housing_program", "string", "If client exited to a housing program, which program was it?", false],
 ["caseworker_exit_housing_program_other", "string", "If other housing program, please specify which one.", false],
 ["caseworker_exit_employed_at_entry", "boolean", "Was resident employed when they entered the shelter?", false],
 ["caseworker_exit_hourly_rate_at_entry", "float", "Hourly rate at employment", false],
 ["caseworker_exit_employed_at_exit", "boolean", "Is resident employed at time of exit?", false],
 ["caseworker_exit_hourly_rate_at_exit", "float", "Hourly rate at exit", false],
 ["caseworker_exit_gained_employment",
  "integer",
  "Auto populated. If they gained employment in shelter, this should say \"1\" (1 = they gained employment while in shelter, 0 = they either lost employment while in shelter, was employed the entire time in shelter, or never was employed while in shelter. This helps us track how many clients got employment while in shelter.)",
  false],
 ["caseworker_exit_employment_type", "string", "If employed at exit was it", false],
 ["caseworker_exit_income_increase", "float", "How much did income increase while in shelter?", false],
 ["caseworker_exit_savings_at_entry", "float", "How much money did the resident have in savings when entering the shelter?", false],
 ["caseworker_exit_savings_at_exit", "float", "How much money did the resident have in savings when exiting the shelter?", false],
 ["caseworker_exit_savings_difference", "float", "Difference between savings", false],
 ["caseworker_exit_savings_source", "string", "If client left with savings, was it from employment, benefits, both or other? (choose one answer)", false],
 ["caseworker_exit_special_status", "string", "Special status", false],
 ["caseworker_exit_members_in_family", "integer", "How many members in family?", false],
 ["caseworker_exit_number_of_adults", "integer", "Number of adults", false],
 ["caseworker_exit_number_of_children", "integer", "Number of children", false],
 ["caseworker_exit_highest_level_achieved", "string", "Highest level achieved in shelter", false],
 ["caseworker_exit_one_month_follow_up", "string", "One month follow-up", true],
 ["caseworker_exit_referrals_at_follow_up", "string", "Referrals given at follow-up", false]]
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/caseworker_exit_survey_loader.rb [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+      parse_date(row.field_value(ASSESSMENT_DATE_COL))
+    end
+
+    def row_assessment_id(row)
+      response_id = row.field_value(RESPONSE_ID_COL)
+      "caseworker_exit-eto-#{response_id}"
+    end
+
+    def form_definition_identifier
+      'caseworker_exit'
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/caseworker_exit_survey_loader.rb]:
```suggestion
      'caseworker_exit_survey'
```

File: drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/aum_prevention_screening_loader.rb [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    # use the eto response id to construct the custom assessment id
+    def row_assessment_id(row)
+      response_id = row.field_value(RESPONSE_ID_COL)
+      "aum_prevention-eto-#{response_id}"
+    end
+
+    def form_definition_identifier
+      'aum_prevention'
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/aum_prevention_screening_loader.rb]: aum-prevention-screening

**Review by gigxz — COMMENTED** [review]
thanks @ttoomey!  lgtm once keys and identifier match staging

**Comment by ttoomey** [comment]
@gigxz I totally misremembered that we were prefixing the keys with the form definition. Thanks for catching that! 

changes:
* modified the current configs to have the correct prefixes
* corrected the form-definition-identifiers to match the identifiers on staging

**Review by gigxz — APPROVED** [review]
Thanks @ttoomey looks good to me!

**Pull request merged**